backtesting/backtesting.py

```diff
@@ -454,128 +653,248 @@ def __init__(self, *, data, cash, commission, margin, trade_on_close, length):
+                if not is_limit_hit or is_limit_hit_before_stop:
+                    continue
+
+                # stop_price, if set, was hit within this bar
+                price = (min(open, order.limit, stop_price or np.inf)
+                         if order.is_long else
+                         max(open, order.limit, stop_price or -np.inf))
+            else:
+                # Market-if-touched / market order
+                price = prev_close if self._trade_on_close else open
+                price = (max(price, stop_price or -np.inf)
+                         if order.is_long else
+                         min(price, stop_price or np.inf))
+
+            # Determine entry/exit bar index
+            is_market_order = not order.limit and not stop_price
+            time_index = (self._i - 1) if is_market_order and self._trade_on_close else self._i
+
+            # If order is a SL/TP order, it should close an existing trade it was contingent upon
+            if order.parent_trade:
+                trade = order.parent_trade
+                # If this trade isn't already closed (e.g. on multiple `trade.close(.5)` calls)
+                if trade in self.trades:
+                    self._reduce_trade(trade, price, order.size, time_index)
+                    assert order.size != -trade.size or trade not in self.trades
+                if order in (trade._sl_order,
+                             trade._tp_order):
+                    assert order.size == -trade.size
+                    assert order not in self.orders  # Removed when trade was closed
+                else:
+                    # It's a trade.close() order, now done
+                    assert abs(trade.size) >= abs(order.size) >= 1
+                    self.orders.remove(order)
+                continue
+
+            # Else this is a stand-alone trade
+
+            # Adjust price to include commission (or bid-ask spread).
+            # In long positions, the adjusted price is a fraction higher, and vice versa.
+            adjusted_price = price * (1 + (np.sign(order.size) * self._commission))
+
+            # If order size was specified proportionally,
+            # precompute true size in units, accounting for margin and spead/commissions
+            size = order.size
+            if -1 < size < 1:
+                size = np.sign(size) * int((self.margin_available * self._leverage * abs(size))
+                                           // adjusted_price)
+                # Not enough cash/margin even for a single unit
+                if not size:
+                    continue
+            assert size == round(size)
+            need_size = int(size)
+
+            # Fill position by FIFO closing/reducing existing opposite-facing trades.
+            # Existing trades are closed at unadjusted price, because the adjustment
+            # was already made when buying.
+            for trade in list(self.trades):
```

Comment:
> If I wanted to disable the NFA compliant FIFO functionality, would that be as simple as commenting out lines 824 to 842 in backtesting.py?

@qacollective Yes, I believe so.

I was under the impression the rule was enforced for all _retail_ traders. It's at least consistent with what I see on Oanda Europe. Maybe only applies to forex. We should definitely make this optional. :+1:

Reply:
@kernc I'm only just learning of this too, but from my current understanding it appears that hedging Forex is illegal in the USA (the land of the 'free' :roll_eyes:) but legal almost everywhere else (https://www.forexbonuses.org/trading-methods/hedging/).

There are even broker review websites which help you find brokers who don't comply with FIFO: https://www.benzinga.com/money/hedging-forex/ :laughing: